url2b9f720f64_jfimbett_QBnB_tgz-pJ8-tests_TestJ8/src/model/Model.java

```diff
@@ -0,0 +1,988 @@
+				}
+				else{
+					p("+"+d+""+namesof.get(j));
+				}
+		    	
+			}
+		    if(c.getType()==0){
+		    	p("="+c.getRs());
+		    }
+		    else if(c.getType()==1){
+		    	p("<="+c.getRs());
+		    }
+		    else{
+		    	p(">="+c.getRs());
+		    }
+		    pl();
+		}
+		
+	}
+	 public boolean isSparse(){
+		 boolean sparse=false;
+		 int count=0;
+		 for (int i = 0; i < A.length; i++) {
+			for (int j = 0; j < A[0].length; j++) {
+				if(A[i][j]!=0){
+					count++;
+				}
+			}
+		}
+		 if(count/(A.length*A[0].length)<0.3){
+			 sparse=true;
+		 }
+		 return sparse;
+	 }
+	/**
+	 * Solve LP Problem
+	 * @param type: 1 If Maximization, 0 if Minimization
+	 */
+	public void solveLP(int type){
+		
+		if(type==1){
+			
+			if(isSparse()){
+				System.out.println("Matrix is sparse");
+			}
+			ArrayList<String> v= variableNames();
+			I= new String[v.size()];
+			v.toArray(I);
+			double[] d=changeSign(c);
+			simplex=new SimplexAlgorithm(A, b, d , I, U, L, print);
+			this.setFo(simplex.primalfo());
+			int status=simplex.getStatus();
+			if(status==3){
+				this.feasible=false;
+			}
+			
+		}
+		else{
+			I=varNames();
+			simplex=new SimplexAlgorithm(A, b, c, I, U, L, print);
+			this.setFo(simplex.primalfo());
+			int status=simplex.getStatus();
+			if(status==3){
+				this.feasible=false;
+			}
+		}
+		
+	}
+
+	/**
+	 * Returns the names of the variables
+	 * @return A vector with the different name of variables
+	 */
+	public String[] varNames(){
+		String[] names= new String[variableNames().size()];
+		for (int i = 0; i < names.length; i++) {
+			names[i]=variableNames().get(i);
+			}
+		return names;
+	}
+	
+	/**
+	 * Changes the sign of each element in a vector
+	 * @param cc: The original vector
+	 * @return -cc
+	 */
+	public double[] changeSign(double[] cc){
+		double[] a=new double[cc.length];
+		for (int i = 0; i < a.length; i++) {
+			a[i]=-cc[i];
+		}
+		return a;
+		
+	}
+
+ /**
+  * Update variables after adding the constraints. 
+  */
+	public void update(){
+		
+		updateConstraintsE();
+		updateBounds();
+		
+	}
+	
+	public void changeLBoundToVariable(String varName, double nlb){
+		QBnBVariable v= findVarByName(varName);
+		v.setLb(nlb);
+	}
+	
+
+	public void changeUBoundToVariable(String varName, double nub){
+		QBnBVariable v= findVarByName(varName);
+		v.setUb(nub);
+	}
+	public void updateBounds(){
+		int n=variables.size();
+		U= new double[n];
+		L= new double[n];
+		
+		for (int i = 0; i < n; i++) {
+			QBnBVariable v= variables.get(i);
+			U[i]=v.getUb();
+			L[i]=v.getLb();
+		}
+		
+	}
+	
+	/**
+	 * Restore the solutions to every variable in the model. 
+	 */
+	public void restoreSolutions(){
+		//Values for variables and restrictions
+		double[] xb=this.getSimplex().getXb();
+		String[] Ib= this.getSimplex().getIb();
+		for (int i = 0; i < xb.length; i++) {
+			String name=Ib[i];
+				int pos=posVarByName(name);
+				if(pos!=-1){
+					variables.get(pos).setValue(xb[i]);
+				}
+				
+		}	
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 332:339, 343:350.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.